ee/agent/storage/bbolt/backup.go

```diff
@@ -0,0 +1,185 @@
+package agentbbolt
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/kolide/launcher/ee/agent/types"
+	"go.etcd.io/bbolt"
+)
+
+const (
+	backupInitialDelay = 10 * time.Minute
+	backupInterval     = 1 * time.Hour
+)
+
+// databaseBackupSaver periodically takes backups of launcher.db.
+type databaseBackupSaver struct {
+	knapsack    types.Knapsack
+	slogger     *slog.Logger
+	interrupt   chan struct{}
+	interrupted bool
+}
+
+func NewDatabaseBackupSaver(k types.Knapsack) *databaseBackupSaver {
+	return &databaseBackupSaver{
+		knapsack:  k,
+		slogger:   k.Slogger().With("component", "database_backup_saver"),
+		interrupt: make(chan struct{}, 1),
+	}
+}
+
+func (d *databaseBackupSaver) Execute() error {
+	// Wait a little bit after startup before taking first backup, to allow for enrollment
+	select {
+	case <-d.interrupt:
+		d.slogger.Log(context.TODO(), slog.LevelDebug,
+			"received external interrupt during initial delay, stopping",
+		)
+		return nil
+	case <-time.After(backupInitialDelay):
+		break
+	}
+
+	// Take periodic backups
+	ticker := time.NewTicker(backupInterval)
+	defer ticker.Stop()
+	for {
+		if err := d.backupDb(); err != nil {
+			d.slogger.Log(context.TODO(), slog.LevelWarn,
+				"could not perform periodic database backup",
+				"err", err,
+			)
+		}
+
+		select {
+		case <-ticker.C:
+			continue
+		case <-d.interrupt:
+			d.slogger.Log(context.TODO(), slog.LevelDebug,
+				"interrupt received, exiting execute loop",
+			)
+			return nil
+		}
+	}
+}
+
+func (d *databaseBackupSaver) Interrupt(_ error) {
+	// Only perform shutdown tasks on first call to interrupt -- no need to repeat on potential extra calls.
+	if d.interrupted {
+		return
+	}
+	d.interrupted = true
+
+	d.interrupt <- struct{}{}
+}
+
+func (d *databaseBackupSaver) backupDb() error {
+	// Take backup -- it's fine to just overwrite previous backups
+	backupLocation := BackupLauncherDbLocation(d.knapsack.RootDirectory())
+	if err := d.knapsack.BboltDB().View(func(tx *bbolt.Tx) error {
```

Comment:
I'm not sure we need to do this now but this is adding another direct bbolt invocation and I know we were trying to move away from that - it gets a little weird because this is a full db copy but I think we could add something like a BackupSaver interface to types/keyvalue_store.go and just define this bit as e.g. `BackupFull` on `bboltKeyValueStore`. then call from `k.ConfigStore`, etc. mostly just thinking out loud, happy to think more and come back to this later

Reply:
Yeah, I keep struggling with this. It seems incorrect to have a full-db operation `BackupFull` attached to a partial-db object (store/bucket `bboltKeyValueStore`) -- but maybe that's unavoidable and the lesser evil if we really want to avoid the direct bbolt invocation?

Reply:
yeah i have not been able to think of a clean solution there, maybe it's best to avoid complicating this for now anyway. there is no major benefit right now

Reply:
Seems fine for now. I think one path might be to make `Backup(ctx, path)` part of the store interface, and then call it from knapsack or something